client_auth.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2016 Mender Software AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/mendersoftware/log"
+	"github.com/pkg/errors"
+)
+
+type AuthGetter interface {
+	Request(identity, deviceID string) (interface{}, error)
+}
+
+// source of authorization data
+type AuthDataProvider interface {
+	// returns device public key
+	Pubkey() (string, error)
+	// returns tenant token
+	Token() (string, error)
+	// returns identity data
+	IdentityData() (interface{}, error)
+}
+
+type AuthClient struct {
+	client *http.Client
+}
+
+func NewAuthClient(conf httpsClientConfig) (*AuthClient, error) {
+	client, err := NewHttpClient(conf)
+	if client == nil {
+		return nil, errors.Wrap(err, "failed to create auth client HTTP client")
+	}
+
+	au := AuthClient{
+		client: client,
+	}
+	return &au, nil
+}
+
+func (u *AuthClient) Request(server string, datasrc AuthDataProvider) (interface{}, error) {
+
+	req, err := makeAuthRequest(server, datasrc)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to build auth request")
+	}
+
+	log.Debugf("making authorization request to server %s with req: %s", server, req)
+	rsp, err := u.client.Do(req)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to execute request")
+	}
+	defer rsp.Body.Close()
+
+	log.Debugf("receive response data")
+	data, err := ioutil.ReadAll(rsp.Body)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to receive response data")
+	}
+
+	// validate response
+	log.Debugf("received response data %v", data)
+
+	return data, nil
+}
+
+type authReqData struct {
+	IdData      interface{} `json:"id_data"`
+	TenantToken string      `json:"tenant_token"`
+	Pubkey      string      `json:"pubkey"`
+}
+
+func makeAuthRequest(server string, datasrc AuthDataProvider) (*http.Request, error) {
+	url := buildURL(server) + "/v0.1/auth_requests"
```

Comment:
Updated to send requests to `/api/0.0.1/authorization/auth_requests`. This will not work with current API gateway setup (actually the previous one didn't work either :) ).